upgrade exporter to v3

jedi-of-the-sea opened the pull request:
Upgrade the exporter addon for Masterportal v3. 

I applied the same notes from the importer addon PR here, too.

jedi-of-the-sea committed:
upgrade exporter to v3

jedi-of-the-sea commented:
@dnlkoch pinging here as a kindly reminder just in case this PR got lost due to other work.

jansule commented on exporter/components/ExporterAddon.vue:
```diff
@@ -0,0 +1,152 @@
+import FlatButton from "@shared/modules/buttons/components/FlatButton.vue";
+
+export default {
+    // eslint-disable-next-line vue/multi-word-component-names, vue/match-component-file-name
```
Is there a specific reason for not naming the file `Exporter.vue` and thereby avoiding the need of adding this setting here?

jedi-of-the-sea replied:
I cannot find an error in your config. I tried to recreate your error, but it did work for me.

The reason for the naming was that the addon was named "ExporterAddon" before - because the addon for the Masterportal v2 was also called ExporterAddon. But in the PR for the importer addon it was remarked that I should rename it, so I did the same here accordingly.

jansule replied:
Thanks for the clarification. Then it is ok as is :+1:

jansule commented on README.md:
```diff
@@ -17,7 +17,8 @@ Please check the `README` or `doc` files within the respective addon folder.
 | simpleLineChart     | Renders data points on simple chart.js component      | v3.8.0       | [FOSSGIS WS](https://github.com/terrestris/masterportal-ws)  |
 | tourGuide     | This addon enables a customised tour through the masterportal. Elements can be highlighted and provided with descriptions and graphics. Technically, it is basically a wrapper for [intro.js](https://introjs.com/) with minor adaptions to fit the masterportal design.      | v3.8.0       | [Meckenheim Geoportal](https://geoportal.meckenheim.de)  |
-| Importer (v3)           | Import common geodata formats and OGC services        | v.3.13.0    |      |
+| importer (v3)       | Import common geodata formats and OGC services        | v.3.13.0    |      |
```
```suggestion
| importer (v3)       | Import common geodata formats and OGC services        | v3.13.0    |      |
```

jansule commented on README.md:
```diff
@@ -17,7 +17,8 @@ Please check the `README` or `doc` files within the respective addon folder.
 | tourGuide     | This addon enables a customised tour through the masterportal. Elements can be highlighted and provided with descriptions and graphics. Technically, it is basically a wrapper for [intro.js](https://introjs.com/) with minor adaptions to fit the masterportal design.      | v3.8.0       | [Meckenheim Geoportal](https://geoportal.meckenheim.de)  |
-| Importer (v3)           | Import common geodata formats and OGC services        | v.3.13.0    |      |
+| importer (v3)       | Import common geodata formats and OGC services        | v.3.13.0    |      |
+| exporter (v3)       | Export WFS and vector layer to common geodata formats | v.3.13.0    |      |
```
```suggestion
| exporter (v3)       | Export WFS and vector layer to common geodata formats | v3.13.0    |      |
```

jansule commented on exporter/package.json:
Could you also add a package-lock.json?

jedi-of-the-sea replied:
done
f59274dfab0b1d1d6180c57597a6cc76655bb0ce

jansule replied:
Great thanks!

jansule reviewed (commented):
Thanks @jedi-of-the-sea, and sorry for the late reply. I just have some minor remarks.

I'm also experiencing some problems setting up the addon, in order to manually  test your changes. I'm probably doing something wrong, but you might be able to help me out here:

- I checked out masterportal in version `3.13.0` and your branch of masterportal-addons
- I copied the addon into the addons folder and created a `addonsConf.json`
- I added the `exporter` to `config.js` and `config.json`
- The addon does not load and only shows me empty content

I'm running the code with node `20.12.2` and getting following error in the browser:

```
Error: Cannot find module './exporter/index.js'
    at eval (eval at ./addons lazy recursive ^\.\/.*$ include: addons[\\\/].*[\\\/]index.js$ exclude: (master/node_modules)|(.+unittests.)|(.+test.)+ (https://localhost:9001/build/js/masterportal.js:211:1), <anonymous>:5:11)
```

Here is my directory structure (irrelevant parts omitted):

```
masterportal
├── addons
│   ├── addonsConf.json
│   ├── exporter
│   │   ├── components
│   │   │   ├── ExporterAddon.vue
│   │   │   ├── LayerDownloadOptions.vue
│   │   │   ├── LayerSelection.vue
│   │   │   └── RadioButton.vue
│   │   ├── constants
│   │   │   ├── exportformats.js
│   │   │   ├── geoPackageDataTypes.js
│   │   │   ├── layertypes.js
│   │   │   ├── steps.js
│   │   │   └── workflows.js
│   │   ├── doc
│   │   │   └── config.json.md
│   │   ├── index.js
│   │   ├── locales
│   │   │   ├── de
│   │   │   │   └── additional.json
│   │   │   └── en
│   │   │       └── additional.json
│   │   ├── package.json
│   │   ├── package-lock.json
│   │   ├── README.md
│   │   ├── store
│   │   │   ├── actionsExporterAddon.js
│   │   │   ├── ExporterAddon.js
│   │   │   ├── gettersExporterAddon.js
│   │   │   ├── mutationsExporterAddon.js
│   │   │   └── stateExporterAddon.js
│   │   ├── tests
│   │   │   ├── end2end
│   │   │   │   └── ExporterAddon.e2e.js
│   │   │   └── unit
│   │   │       ├── components
│   │   │       │   └── ExporterAddon.spec.js
│   │   │       └── store
│   │   │           ├── actionsExporterAddon.spec.js
│   │   │           ├── gettersExporterAddon.spec.js
│   │   │           └── mutationsExporterAddon.spec.js
│   │   └── utils
│   │       ├── download.js
│   │       ├── layer.js
│   │       └── resetStep.js
│   └── readme.txt
├── package.json
├── package-lock.json
├── portal
│   └── master
│       ├── config.js
│       ├── config.json
│       ├── index.html
│       ├── resources
│       │   ├── customMenuElementContent.html
│       │   ├── example_geojson.json
│       │   ├── newsFeedPortalAlerts.json
│       │   ├── vectorTile.png
│       │   └── webVector.png
│       └── webgl_assets
│           └── dipas.test.geojson
```
addonsConf.json
```json
{
  "exporter": {
    "type": "tool"
  }
}
```
config.js (irrelevant parts omitted)
```js
const Config = {
    addons: ["exporter", "bildungsatlas", "dataTable", "dipas", "exampleSearch", "populationRequest", "schulinfo", "solaratlas", "sturmflut", "trafficCount", "trinkwasser", "verkehrsstaerken"],
```

config.json (irrelevant parts omitted)
```json
 "mainMenu": {
      "expanded": true,
      "title": {
        "text": "Master",
        "logo": "https://geodienste.hamburg.de/lgv-config/img/hh-logo.png",
        "link": "https://geoinfo.hamburg.de",
        "toolTip": "Landesbetrieb Geoinformation und Vermessung"
      },
      "sections": [
        [
          {
            "type": "exporter"
          },
```

jedi-of-the-sea committed:
Update README.md

Co-authored-by: Jan Suleiman <jansule@users.noreply.github.com>

jedi-of-the-sea committed:
Update README.md

Co-authored-by: Jan Suleiman <jansule@users.noreply.github.com>

jedi-of-the-sea committed:
add package-lock.json

jedi-of-the-sea committed:
Merge branch 'upgrade-exporter-to-v3' of https://github.com/jedi-of-the-sea/masterportal-addons into upgrade-exporter-to-v3

jansule commented:
Thanks for addressing the remarks @jedi-of-the-sea!

I was finally able to get this running and it looks good so far!

There is just one thing, that I noticed: The exporter currently seems to only respect layers that have been visible/active when loading the application. Layers that have been added afterwards to the layer tree (any WFS, but also drawings created with draw_old) are not listed in the exporter. Do you know, if that behavior already existed in v2?

jedi-of-the-sea commented:
> Thanks for addressing the remarks @jedi-of-the-sea!
> 
> I was finally able to get this running and it looks good so far!
> 
> There is just one thing, that I noticed: The exporter currently seems to only respect layers that have been visible/active when loading the application. Layers that have been added afterwards to the layer tree (any WFS, but also drawings created with draw_old) are not listed in the exporter. Do you know, if that behavior already existed in v2?

That's weird.. the exporter should be able to find invisible layers as well as added layers (wfs imported by importer addon for instance).. I just tried that in my set-up and it works.. but maybe I misunderstood and you meant that you have to re-open the addon for this to work? The exporter as it is now does not dynamically add new layers for download if they are added while the exporter addon is opened.

I'm not sure if the v2 addon was able to handle drawings with draw_old, but I searched the code for this type and did not find any handling for layers with type "draw_old".

jansule commented:
> That's weird.. the exporter should be able to find invisible layers as well as added layers (wfs imported by importer addon for instance).. I just tried that in my set-up and it works.. but maybe I misunderstood and you meant that you have to re-open the addon for this to work? The exporter as it is now does not dynamically add new layers for download if they are added while the exporter addon is opened.

Now it's working for me too, not sure what was the reason for the previous behavior...

> I'm not sure if the v2 addon was able to handle drawings with draw_old, but I searched the code for this type and did not find any handling for layers with type "draw_old".

In v2 we were able to also export layers created by the draw tool. But apparently that changed. The code now uses `Importierte Layer` as label for that section, which also seems to work with imported vector layers. So I'm ok with that.

jansule reviewed (approved):
Thanks for putting all the efforts into this upgrade! This can be merged now
